### ISSUE TITLE
AverageReturn  metrics doesn't calculate the correct return?

### ISSUE THREAD

**Issue by WangMengqi32C** [tensorflow/agents#573]
Dear all,
I'm using tf_agents.metrics.tf_metrics.AverageReturnMetric like the example in PPO algorithm and I found the average return, maximal return and minimal return plotted in tensorboard are different from the one I calculate without metrics. They are on a completely different scale.

Here is the way I define the metrics:
train_metrics = [tf_metrics.NumberOfEpisodes(),
        tf_metrics.EnvironmentSteps(),
        tf_metrics.AverageReturnMetric(batch_size=num_parallel_environments),
        tf_metrics.AverageEpisodeLengthMetric(batch_size=num_parallel_environments),
        tf_metrics.MaxReturnMetric(batch_size=num_parallel_environments),
        tf_metrics.MinReturnMetric(batch_size=num_parallel_environments)
        ]
eval_summary_writer = tf.compat.v2.summary.create_file_writer(eval_dir,flush_millis=10000)
eval_metrics = [tf_metrics.NumberOfEpisodes(),
        tf_metrics.EnvironmentSteps(),
        tf_metrics.AverageReturnMetric(buffer_size=collect_episodes_per_iteration),#buffer_size=collect_episodes_per_iteration
        tf_metrics.AverageEpisodeLengthMetric(buffer_size=collect_episodes_per_iteration),
        ]
train_checkpointer = common.Checkpointer(ckpt_dir=train_dir,
                                          agent=agent, 
                                          policy=agent.policy,
                                          replay_buffer=replay_buffer,
                                          global_step=global_step, 
                                          metrics=metric_utils.MetricsGroup(train_metrics, 'train_metrics'))
policy_checkpointer = common.Checkpointer(ckpt_dir=os.path.join(train_dir, 'policy'),
                                          policy=agent.policy,
                                          global_step=global_step)

train_checkpointer.initialize_or_restore()

collect_driver = dynamic_episode_driver.DynamicEpisodeDriver(train_env, 
                                                            agent.collect_policy,
                                                            observers=[replay_buffer.add_batch]+ train_metrics,
                                                            num_episodes=collect_episodes_per_iteration)

And here is how I calculate the average return:
def compute_avg_return(environment, policy, num_episodes=10):

  total_return = 0.0
  for _ in range(num_episodes):

    time_step = environment.reset()
    episode_return = 0.0
    #print(environment.batch_size)
    state = policy.get_initial_state(environment.batch_size)
    #print(time_step)
    while not time_step.is_last():
      #print(state)
      #print(time_step)
      action_step = policy.action(time_step,policy_state=state)
      #print(action_step)
      time_step = environment.step(action_step.action)
      #print(time_step)
      episode_return += time_step.reward
    total_return += episode_return

  avg_return = total_return / num_episodes
  return avg_return.numpy()[0] 

In the figure, we can see the return is always less than 1,  but with the same environment, it's about 1000 by my method. I think the number from my method is more trustworthy because, in every step, the agent can obtain a reward in the range of 0 to 1, and it takes 2000 steps in every episode.
![metric_returns_wrong](https://user-images.githubusercontent.com/37926759/110813676-0458f000-8289-11eb-9115-f66aa68c48e2.png

I have been stuck here for a while, really hope someone can help me!

Thank you in advance!
Best regards,
Mengqi

**Comment by ebrevdo** [comment]
@summer-yue any ideas?

**Comment by summer-yue** [comment]
Thanks for your question. Before I try reproducing this with a standard environment, it'd be very helpful if you can provide some more information. I see that the screenshot you linked when the metrics are of the wrong scale. Seems like this corresponds to the `train_metrics`? Are you seeing the same problem for the eval metrics? (This can help me narrow down whether it's the batch_size argument that's causing issues)

And what did you set for the values of `collect_episodes_per_iteration` and `num_parallel_environments`?

**Comment by WangMengqi32C** [comment]
@summer-yue  @ebrevdo   
hi all, thank you for your help.
 I don't see the same problem for eval metrics. I think the reason is the policy I used. I used agent.collect_policy to collect the data in replay_butter, and I use agent.policy to update the eval metric. If I use greedy_policy.GreedyPolicy(agent.collect_policy) to collect data, then the average return from train metric is same as the average return from the function I used to calculate the average return.

I tried the mountain car continuous environment (from open AI gym). There wasn't a big difference between agent.collect_policy and agent.policy. Perhaps it's because my environment is too sensitive to the slight change of the policy(?) I don't know how agent.collect_policy is different from agent.policy in the PPO class.(from the algorithm, by my understanding it should be same??) I'm working on find it out.

And I set collect_episodes_per_iteration=20, num_parallel_environments=20. 

here is how i defined the driver:
collect_driver = dynamic_episode_driver.DynamicEpisodeDriver(train_env, 
                                                            greedy_policy.GreedyPolicy(agent.collect_policy),
                                                            observers=[replay_buffer.add_batch]+ train_metrics,
                                                            num_episodes=collect_episodes_per_iteration)
here is how I use eval metrics:
eval_summary_writer = tf.compat.v2.summary.create_file_writer(eval_dir,flush_millis=10000)
eval_metrics = [tf_metrics.NumberOfEpisodes(),
        tf_metrics.EnvironmentSteps(),
        tf_metrics.AverageReturnMetric(buffer_size=collect_episodes_per_iteration),#buffer_size=collect_episodes_per_iteration
        tf_metrics.AverageEpisodeLengthMetric(buffer_size=collect_episodes_per_iteration),
        ]
results=metric_utils.eager_compute(
                eval_metrics,
                eval_env,
                agent.policy,
                num_episodes=num_eval_episodes,
                train_step=global_step,
                summary_writer=eval_summary_writer,
                summary_prefix='Metrics',
             )

Thank you for your help again!
Mengqi

**Comment by summer-yue** [comment]
`agent.policy` is just the greedy version of`agent.collect_policy` in PPO. You can find the definitions here:  https://github.com/tensorflow/agents/blob/master/tf_agents/agents/ppo/ppo_agent.py#L345

The greedy policy would always take the action with the highest probability, whereas the non greedy counterpart may sample other policies as well. There are situations when they can get drastically different returns.

One way to verify this is to print out `policy.distribution(timestep)` and see what the action probabilities looks like, and analyze if the greedy action gets drastically better returns than then other ones.

**Comment by ebrevdo** [comment]
@summer-yue would you consider this issue resolved?

**Comment by summer-yue** [comment]
Resolving the issue as it's working as intended.

**Comment by WangMengqi32C** [comment]
Hi Summer and ebrevdo,

Sorry for late replay, i did check this email.

Your suggestions help a lot, thank you for that.

Hope you have a nice weekends,
Mengqi

发件人：Summer Yue ***@***.***>
发送日期：2021-05-07 05:42:53
收件人：tensorflow/agents ***@***.***>
抄送人：Mengqi Wang ***@***.***>,Author ***@***.***>
主题：Re: [tensorflow/agents] AverageReturn metrics doesn't calculate the correct return? (#573)

Resolving the issue as it's working as intended.
